### PR TITLE
Add support for creating a memcached instance that uses a specific IP address range

### PR DESCRIPTION
**Pull request by kshitijjaiswal2**
Add support for creating a memcached instance that uses a specific IP address range.
Github issue - fixes https://github.com/hashicorp/terraform-provider-google/issues/16819

```release-note:enhancement
memcache: added `reserved_ip_range_id` field to `google_memcache_instance` resource
```

### CODE REVIEW

**Comment by kshitijjaiswal2** [comment]
@roaks3 can you please review this?

**Review by roaks3 — CHANGES_REQUESTED** [review]
I had a few questions about the config, but once that is resolved, the field should be added to at least one test. In this case, that would most likely be `testAccMemcacheInstance_update()` or `testAccMemcacheInstance_update2()`.

File: mmv1/products/memcache/Instance.yaml
Comment (kshitijjaiswal2) [reply mmv1/products/memcache/Instance.yaml]: @roaks3 thanks for making things clear i have added the suggested changes we'll only keep ignore_read as the value is used for inital creation request only.

**Review by roaks3 — CHANGES_REQUESTED** [review]
Code LGTM, but could you also please add this to at least one test as mentioned in my previous comment?

File: mmv1/products/memcache/Instance.yaml
Comment (kshitijjaiswal2) [reply mmv1/products/memcache/Instance.yaml]: @roaks3 i have added the field to testAccMemcacheInstance_update().

**Comment by kshitijjaiswal2** [comment]
@roaks3 is it fine?

**Review by roaks3 — APPROVED** [review]
Yep, LGTM assuming tests pass

**Comment by kshitijjaiswal2** [comment]
@roaks3 can you please merge the code changes if they is fine?

**Comment by roaks3** [comment]
Hmm, I will have to check into this. It appears that for some reason the tests didn't run in CI (or they didn't report the results).

**Comment by roaks3** [comment]
Ah ok, this is due to an issue that was introduced at the end of December but has since been resolved.

@kshitijjaiswal2 could you please rebase this PR so that it is caught up with the latest changes from main? That should allow the automated tests to run (and hopefully pass so that we can merge)

**Comment by roaks3** [comment]
It looks like the API is returning `"authorizedNetwork": "projects/ci-test-project-188019/global/networks/tf-bootstrap-net-memcache-instance-update-1"`, which doesn't match the request structure (`reservedIpRangeId: ...`), so it is not able to resolve state properly after that point.

This is non-standard API behavior that we tend to not encourage. To make this work, you may need to use `ignore_read: true` so that the API does not need to include `reservedIpRangeId` in the response, but this treats the field more like an initial setting only. I believe you can leave `authorizedNetwork` as-is, but it's possible you will need to do something to handle the new value being populated, with something like ignore_read or a diff suppress.

**Comment by kshitijjaiswal2** [comment]
@roaks3 could you please share the debug logs or error message with me, i don't have access to it, if possible can you grant me access to these logs?

**Comment by roaks3** [comment]
Sure, here are the relevant request/responses from the debug log:
```
---[ REQUEST ]---------------------------------------
POST /v1beta2/projects/ci-test-project-188019/locations/us-central1/instances?alt=json&instanceId=tf-test-6754803316342026879 HTTP/1.1
Host: memcache.googleapis.com
User-Agent: Terraform/1.2.5 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google-beta/acc
Content-Length: 356
Content-Type: application/json
Accept-Encoding: gzip

{
 "authorizedNetwork": "projects/ci-test-project-188019/global/networks/tf-bootstrap-net-memcache-instance-update-1",
 "memcacheVersion": "MEMCACHE_1_5",
 "nodeConfig": {
  "cpuCount": 1,
  "memorySizeMb": 1024
 },
 "nodeCount": 1,
 "parameters": {
  "params": {
   "listen-backlog": "2048",
   "max-item-size": "8388608"
  }
 },
 "reservedIpRangeId": [
  "tf-bootstrap-addr-memcache-instance-update-1"
 ]
}

-----------------------------------------------------
2024/01/08 16:37:52 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Mon, 08 Jan 2024 16:37:52 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/ci-test-project-188019/locations/us-central1/operations/operation-1704731872085-60e71d1ac4770-bfe4185a-e0e7bacc",
  "metadata": {
    "@type": "type.googleapis.com/google.cloud.memcache.v1beta2.OperationMetadata",
    "createTime": "2024-01-08T16:37:52.591934977Z",
    "target": "projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879",
    "verb": "create",
    "cancelRequested": false,
    "apiVersion": "v1beta2"
  },
  "done": false
}
...
---[ REQUEST ]---------------------------------------
GET /v1beta2/projects/ci-test-project-188019/locations/us-central1/operations/operation-1704731872085-60e71d1ac4770-bfe4185a-e0e7bacc?alt=json HTTP/1.1
Host: memcache.googleapis.com
User-Agent: Terraform/1.2.5 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google-beta/acc
Content-Type: application/json
Accept-Encoding: gzip


-----------------------------------------------------
2024/01/08 16:42:38 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Mon, 08 Jan 2024 16:42:38 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/ci-test-project-188019/locations/us-central1/operations/operation-1704731872085-60e71d1ac4770-bfe4185a-e0e7bacc",
  "metadata": {
    "@type": "type.googleapis.com/google.cloud.memcache.v1beta2.OperationMetadata",
    "createTime": "2024-01-08T16:37:52.591934977Z",
    "endTime": "2024-01-08T16:42:37.855701774Z",
    "target": "projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879",
    "verb": "create",
    "cancelRequested": false,
    "apiVersion": "v1beta2"
  },
  "done": true,
  "response": {
    "@type": "type.googleapis.com/google.cloud.memcache.v1beta2.Instance",
    "name": "projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879",
    "authorizedNetwork": "projects/ci-test-project-188019/global/networks/tf-bootstrap-net-memcache-instance-update-1",
    "zones": [
      "us-central1-a",
      "us-central1-b",
      "us-central1-c",
"us-central1-f"
    ],
    "nodeCount": 1,
    "nodeConfig": {
      "cpuCount": 1,
      "memorySizeMb": 1024
    },
    "memcacheVersion": "MEMCACHE_1_5",
    "parameters": {
      "id": "-1788763201",
      "params": {
        "listen-backlog": "2048",
        "max-item-size": "8388608"
      }
    },
    "memcacheNodes": [
      {
        "nodeId": "node-f-1",
        "zone": "us-central1-f",
        "state": "READY",
        "host": "10.236.0.4",
        "port": 11211,
        "parameters": {
          "id": "-1788763201",
          "params": {
            "listen-backlog": "2048",
            "max-item-size": "8388608"
          }
        },
        "memcacheVersion": "MEMCACHE_1_5",
        "memcacheFullVersion": "memcached-1.5.16"
      }
    ],
    "createTime": "2024-01-08T16:37:52.588213407Z",
    "updateTime": "2024-01-08T16:37:52.588213407Z",
    "state": "READY",
    "memcacheFullVersion": "memcached-1.5.16",
    "discoveryEndpoint": "10.236.0.3:11211"
  }
}

-----------------------------------------------------
2024/01/08 16:42:38 [DEBUG] Retry Transport: Stopping retries, last request was successful
2024/01/08 16:42:38 [DEBUG] Retry Transport: Returning after 1 attempts
2024/01/08 16:42:38 [DEBUG] Got done: true while polling for operation projects/ci-test-project-188019/locations/us-central1/operations/operation-1704731872085-60e71d1ac4770-bfe4185a-e0e7bacc's status
2024/01/08 16:42:38 [DEBUG] Finished creating Instance "projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879": map[string]interface {}{"done":false, "metadata":map[string]interface {}{"@type":"type.googleapis.com/google.cloud.memcache.v1beta2.OperationMetadata", "apiVersion":"v1beta2", "cancelRequested":false, "createTime":"2024-01-08T16:37:52.591934977Z", "target":"projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879", "verb":"create"}, "name":"projects/ci-test-project-188019/locations/us-central1/operations/operation-1704731872085-60e71d1ac4770-bfe4185a-e0e7bacc"}
2024/01/08 16:42:38 [DEBUG] Waiting for state to become: [success]
2024/01/08 16:42:38 [DEBUG] Retry Transport: starting RoundTrip retry loop
2024/01/08 16:42:38 [DEBUG] Retry Transport: request attempt 0
2024/01/08 16:42:38 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1beta2/projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879?alt=json HTTP/1.1
Host: memcache.googleapis.com
User-Agent: Terraform/1.2.5 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google-beta/acc
Content-Type: application/json
Accept-Encoding: gzip


-----------------------------------------------------
2024/01/08 16:42:38 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Mon, 08 Jan 2024 16:42:38 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/ci-test-project-188019/locations/us-central1/instances/tf-test-6754803316342026879",
  "authorizedNetwork": "projects/ci-test-project-188019/global/networks/tf-bootstrap-net-memcache-instance-update-1",
  "zones": [
    "us-central1-a",
    "us-central1-b",
    "us-central1-c",
"us-central1-f"
  ],
  "nodeCount": 1,
  "nodeConfig": {
    "cpuCount": 1,
    "memorySizeMb": 1024
  },
  "memcacheVersion": "MEMCACHE_1_5",
  "parameters": {
    "id": "-1788763201",
    "params": {
      "listen-backlog": "2048",
      "max-item-size": "8388608"
    }
  },
  "memcacheNodes": [
    {
      "nodeId": "node-f-1",
      "zone": "us-central1-f",
      "state": "READY",
      "host": "10.236.0.4",
      "port": 11211,
      "parameters": {
        "id": "-1788763201",
        "params": {
          "listen-backlog": "2048",
          "max-item-size": "8388608"
        }
      },
      "memcacheVersion": "MEMCACHE_1_5",
      "memcacheFullVersion": "memcached-1.5.16"
    }
  ],
  "createTime": "2024-01-08T16:37:52.588213407Z",
  "updateTime": "2024-01-08T16:42:37.878306603Z",
  "state": "READY",
  "memcacheFullVersion": "memcached-1.5.16",
  "discoveryEndpoint": "10.236.0.3:11211"
}
```

**Comment by roaks3** [comment]
And for reference, this is the error that is ultimately raised:
```
=== RUN   TestAccMemcacheInstance_update
=== PAUSE TestAccMemcacheInstance_update
=== CONT  TestAccMemcacheInstance_update
    vcr_utils.go:152: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) {
        }
        
        
        (map[string]string) (len=2) {
         (string) (len=22) "reserved_ip_range_id.#": (string) (len=1) "1",
         (string) (len=22) "reserved_ip_range_id.0": (string) (len=44) "tf-bootstrap-addr-memcache-instance-update-1"
        }
--- FAIL: TestAccMemcacheInstance_update (499.37s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-google-beta/google-beta/services/memcache	499.420s
FAIL
```

**Comment by kshitijjaiswal2** [comment]
@roaks3 updated with the recommended changes.

**Review by roaks3 — APPROVED** [review]
LGTM, will merge once the tests pass

**Comment by roaks3** [comment]
This is failing during the lint step:
```
Run make lint
sh -c "'/home/runner/work/magic-modules/magic-modules/scripts/gofmtcheck.sh'"
==> Checking that code complies with gofmt requirements...
gofmt needs running on the following files:
./google/services/memcache/resource_memcache_instance_test.go
You can use the command: `make fmt` to reformat code.
make: *** [GNUmakefile:2[6](https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/7630287632/job/20785626737#step:10:7): fmtcheck] Error 1
Error: Process completed with exit code 2.
```

It looks like the indentation needs to be changed after adding `ImportStateVerifyIgnore`.

**Comment by kshitijjaiswal2** [comment]
@roaks3 can you trigger the test once again, or can you tell me how can i trigger the tests?

**Comment by roaks3** [comment]
Yep, kicking them off now. For security reasons, we don't allow contributors to run checks unless they are Google employees, so one of the downsides is that we have to manually click an approval button and pass along any error messages.

**Comment by kshitijjaiswal2** [comment]
@roaks3 i am a google employee :)

**Comment by kshitijjaiswal2** [comment]
@roaks3 the checks have passed, can you merge the PR?

**Comment by roaks3** [comment]
Looks good, merging now. Note that I updated the PR description with the "fixes" keyword to make sure the issue is automatically closed.

Also, since you are a Google employee, make sure you've followed the steps at go/terraform-contribution-guide#before-you-begin to set up your org memberships. Once that is done, you should be able to see the checks automatically run on future PRs.

**Pull request merged**
